gremlin-go/driver/resultSet.go

```diff
@@ -50,25 +50,63 @@ type channelResultSet struct {
 	statusAttributes map[string]interface{}
 	closed           bool
 	err              error
-	mux              sync.Mutex
+	waitSignal       chan bool
+	channelMutex     sync.Mutex
+	waitSignalMutex  sync.Mutex
+}
+
+func (channelResultSet *channelResultSet) sendSignal() {
+	// Lock wait
+	channelResultSet.waitSignalMutex.Lock()
+	defer channelResultSet.waitSignalMutex.Unlock()
+	if channelResultSet.waitSignal != nil {
+		channelResultSet.waitSignal <- true
+		channelResultSet.waitSignal = nil
+	}
 }
 
 func (channelResultSet *channelResultSet) GetError() error {
 	return channelResultSet.err
 }
 
 func (channelResultSet *channelResultSet) IsEmpty() bool {
-	channelResultSet.mux.Lock()
-	defer channelResultSet.mux.Unlock()
-	return channelResultSet.closed && len(channelResultSet.channel) == 0
+	channelResultSet.channelMutex.Lock()
+	// If our channel is empty and we have no data in it, wait for signal that the state has been updated.
+	channelIsEmpty := len(channelResultSet.channel) == 0
+	if !channelIsEmpty {
+		// Channel is not empty.
+		channelResultSet.channelMutex.Unlock()
+		return false
+	} else if channelResultSet.closed {
+		// Channel is empty and closed.
+		return true
```

Comment:
Are you missing a unlock here?
```suggestion
                channelResultSet.channelMutex.Unlock()
		return true
```

Reply:
He defers the unlock in the beginning